server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/dialects/idms/IdmsDialect.java

```diff
@@ -17,69 +17,222 @@
+import java.util.Deque;
+import java.util.LinkedList;
 import java.util.List;
 
 /** Process the text according to the IDMS rules */
 public final class IdmsDialect implements CobolDialect {
   public static final String NAME = "IDMS";
   private final CopybookService copybookService;
-  private final ParseTreeListener treeListener;
   private final MessageService messageService;
 
-  public IdmsDialect(CopybookService copybookService,
-                     ParseTreeListener treeListener,
-                     MessageService messageService) {
+  public IdmsDialect(
+      CopybookService copybookService,
+      MessageService messageService) {
     this.copybookService = copybookService;
-    this.treeListener = treeListener;
     this.messageService = messageService;
   }
 
   /**
    * Gets the name of the dialect
+   *
    * @return the name of the dialect
    */
   @Override
   public String getName() {
     return NAME;
   }
 
+  @Override
+  public List<SyntaxError> extend(DialectProcessingContext context) {
+    List<SyntaxError> errors = new LinkedList<>();
+
+    IdmsDialectVisitor inlineVisitor = new IdmsDialectVisitor(context);
+    IdmsParser.StartRuleContext ruleContext = parseIdms(context.getExtendedSource().getText(), context.getProgramDocumentUri(), errors);
+
+    List<IdmsCopybookDescriptor> cbs = inlineVisitor.visitStartRule(ruleContext);
+    cbs.forEach(cb ->
+            insertIdmsCopybook(
+                    context.getExtendedSource(),
+                    context.getExtendedSource().getMainMap(),
+                    errors, cb,
+                    context.getProgramDocumentUri(),
+                    context.getExtendedSource().getUri(),
+                    context.getCopybookConfig()));
+    return errors;
+  }
+
+  private void insertIdmsCopybook(ExtendedSource extendedSource, DocumentMap currentMap, List<SyntaxError> errors,
+      IdmsCopybookDescriptor cb, String programDocumentUri, String currentUri,
+      CopybookConfig copybookConfig) {
+    insertIdmsCopybook(extendedSource, currentMap, errors, cb, programDocumentUri, currentUri,
+        copybookConfig, new LinkedList<>());
+  }
+
+  private void insertIdmsCopybook(ExtendedSource extendedSource, DocumentMap currentMap, List<SyntaxError> errors,
+                                  IdmsCopybookDescriptor cb, String programDocumentUri, String currentUri,
+                                  CopybookConfig copybookConfig,
+                                  Deque<String> copybookStack) {
+    CopybookModel copybookModel =
+        copybookService.resolve(
+            new CopybookName(cb.getName(), IdmsDialect.NAME),
+            programDocumentUri,
+            currentUri,
+            copybookConfig,
+            true);
+
+    if (copybookModel.getUri() == null || copybookModel.getContent() == null) {
+      errors.add(ErrorHelper.missingCopybooks(messageService, cb.getUsage(), cb.getName()));
+      extendedSource.replace(cb.getStatement().getRange(), "");
```

Comment:
It may be a stupid doubt :) I didn't understand this. Do we make this replacement so that, the rest of the code can be analyzed properly?

Reply:
Yes, the thing is that we remove this code from a source because we "handle" this IDMS statement, so other parsers  (for pure COBOL) will be not confusing wuth unexpected statements